### DISCUSSION TITLE
ST7735 mirrored screen

### DISCUSSION THREAD

**Discussion by nicknugget** [moononournation/Arduino_GFX#352] [Ideas]
I am using the library on a project where it would be helpful if the contents of the display could be mirrored. As far as I have seen, the ST7735 should support such a mode. Is that big effort to support that in one of the future versions with an appropriate method? The partially underlying U8G2 library supports something like that for quite a few OLED displays. 

Thanks for all the work and effort. I really like your library!

**Comment by moononournation** [comment]
The design is actually very easy. The normal rotation is 0-3, the mirrored rotation should be 4-7. It just need someone help to test for each display driver.

**Reply by ZinggJM** [reply]
I would recommend to use a separate method, to keep it consistent with Adafruit_GFX:

```
/**************************************************************************/
/*!
    @brief      Set rotation setting for display
    @param  x   0 thru 3 corresponding to 4 cardinal rotations
*/
/**************************************************************************/
void Adafruit_GFX::setRotation(uint8_t x) {
  rotation = (x & 3);
  switch (rotation) {
  case 0:
  case 2:
    _width = WIDTH;
    _height = HEIGHT;
    break;
  case 1:
  case 3:
    _width = HEIGHT;
    _height = WIDTH;
    break;
  }
}
```

**Reply by ZinggJM** [reply]
There are more design decisions to consider:
- mirror would usually mean flip in x-direction, but mirror for head-up might be flip in y-direction
- rotation is usually done on data input to graphics memory (GRAM, MAD-control).
- mirror could be done on screen output on many TFTs (gate or source control)

**Comment by nicknugget** [comment]
@moononournation 

That would be really great. I'm not sure if every controller your library supports is able to mirror the screen content. So far I only know it for the ST7735.

I would either extend the rotation method as you described or add another method for mirroring. Like the U8G2 lib does.

Since I'm currently using a 0.96" IPS display with a resolution of 80x160px and a ST7735S controller with your library, I can at least give you feedback for that if mirroring works with it.


Update:

After first reaching a dead end mentally, I finally got it.

Rotation and mirroring can be solved using only the MADCTL and RGB-BGR order. So I simply added 4 more modes to the rotation selection. Which allows not only to rotate the screen content, but also to mirror it in each rotation.

At the moment I have simply patched the library for the ST7735. I'll have to look through it at my leisure and see if I haven't overlooked something. But I now have a solution for my project.
